### PR TITLE
Closes #4036: "Try catch" optimisation

### LINKED ISSUE #4036: "Try catch" optimisation
The following code compiles with additional self assignment in catch section when user set exception variable name as _error.

```coffeescript
try something()
catch _error
  console.log _error
```

compiles to:
```javascript
try {
  something();
} catch (_error) {
  // unnecessary operations
  _error = _error;

  console.log(_error);
}
```

Here is extra assignment operation

```coffeescript
try something()
catch error
  console.log error
```

compiles to:
```javascript

try {
  something();
} catch (_error) {
  // extra assignment
  error = _error;
  console.log(error);
}

```
Maybe coffeescript compiler may remove this unnecessary code?

### PR DESCRIPTION
**Pull request by BrunoBernardino**
Let me know if there's something I should be doing differently as this is my first contribution to coffeescript.

I fixed the reported issue where a generated variable could clash with a user-defined one in a try/catch block.

I added a test for a few scenarios with different variable names for a try/catch, to confirm the fix and avoid regressions.

### CODE REVIEW

**Comment by vendethiel** [comment]
What happens if the user has a variable named `__error` then? :-)

**Comment by BrunoBernardino** [comment]
It'll go back to `_error`. Tests prevent it. It's just one variable, both cases won't happen :)

**Comment by vendethiel** [comment]
What I meant to say was that -- instead of hardcoding names, you should use `scope.freeVariable` (like the rest of the compiler does)

You can look at `scope.litcoffee`'s explanation of `freeVariable` to understand why it's more robust.

File: test/scope.coffee [hunk 1]
```diff
@@ -33,6 +33,15 @@ test "catch statements should introduce their argument to scope", ->
...
+test "#4036 catch statements' error variable should not clash with user-defined variable", ->
+  compile = (code) -> CoffeeScript.compile(code, bare: yes).trim().replace(/\s+/g, " ")
+
+  eq "var _error; try { something(); } catch (__error) { _error = __error; console.log(_error); }", compile("try something()\ncatch _error\n  console.log _error")
+
+  eq "var __error; try { something(); } catch (_error) { __error = _error; console.log(__error); }", compile("try something()\ncatch __error\n  console.log __error")
+
+  eq "var error; try { something(); } catch (_error) { error = _error; console.log(error); }", compile("try something()\ncatch error\n  console.log error")
+
 test "loop variable should be accessible after for-of loop", ->
```
Comment (lydell) [inline test/scope.coffee]: Don’t do fragile string comparisons of the output JavaScript. Test that the CoffeeScript works the way intended (just like other tests do). The tests regrading [variables named similar to generated variables](https://github.com/jashkenas/coffeescript/blob/2d1a6fa6ec76b4e721da7e5e4a9c6b90172309ae/test/assignment.coffee#L417-L466) might help.

**Comment by BrunoBernardino** [comment]
Oh, thanks for that, I didn't know! I'll redo this PR with that then. :)

**Comment by vendethiel** [comment]
Great! I'll close this one in the meantime :).

**Comment by BrunoBernardino** [comment]
I thought I could just update this branch? No worries, I'll make another PR and mention this one, then.

**Comment by lydell** [comment]
There’s nothing stopping you from updating this PR. It’ll get re-opened.

**Comment by vendethiel** [comment]
Oh, both works! Sorry.

**Comment by BrunoBernardino** [comment]
@lydell should I add the tests you mentioned, or do those _instead_ of the one I wrote?

**Comment by vendethiel** [comment]
It'd be more robusts to use the new ones instead of the old ones (that's how it currently works)

**Comment by BrunoBernardino** [comment]
This issue does depend on the fact the generated code (not functionality) is "wrong", though...

**Comment by BrunoBernardino** [comment]
I'll test the variable's scope, because it should be the same, and with the previously generated code it wasn't.

**Comment by vendethiel** [comment]
Indeed, please test that instead of the code.

**Comment by BrunoBernardino** [comment]
Done. The test I added, with the previous code, would fail at the second `eq _error, 'foo'` with `AssertionError: Expected undefined to equal foo` :)

All good now.

**Comment by vendethiel** [comment]
LGTM otherwise

File: src/nodes.coffee [hunk 2]
```diff
@@ -1892,15 +1892,17 @@ exports.Try = class Try extends Base
   # is optional, the *catch* is not.
   compileNode: (o) ->
     o.indent  += TAB
+    scope     = o.scope
     tryPart   = @attempt.compileToFragments o, LEVEL_TOP
 
     catchPart = if @recovery
-      placeholder = new Literal '_error'
+      generatedErrorVariableName = scope.freeVariable '_error'
```
Comment (lydell) [inline src/nodes.coffee]: Why not simply `o.scope.freeVariable` and remove `scope = o.scope` (which btw could be written `{scope} = o`) above?

Comment (BrunoBernardino) [inline src/nodes.coffee]: It was just in case the variable was needed for something else. Do you prefer I write it like that?

**Comment by BrunoBernardino** [comment]
Ok. I think it's got all suggestions in now :)

**Comment by jashkenas** [comment]
Nice fix! Thanks for contributing.

**Pull request merged**